### DISCUSSION TITLE
问一下这个内存占用正常吗？

### DISCUSSION THREAD

**Discussion by dudt** [kkkgo/PaoPaoDNS#58] [问答和帮助]
![image](https://github.com/kkkgo/PaoPaoDNS/assets/19968269/eb60f169-f6ee-4bcb-b454-d9897b82a71d)
![image](https://github.com/kkkgo/PaoPaoDNS/assets/19968269/1831dd34-d627-4f30-bbe0-70252a7aa84c)

**Comment by kkkgo** [comment]
内存占用是根据CPU核心数和启动的时候可用内存分配的。吃多少内存取决于你有多少内存。容器启动默认无限制内存，你可以使用docker的相关选项分配资源占用，比如通过-m 选项限制指定容器使用的内存，或者其他管理docker的UI也有相关选项。

**Comment by kkkgo** [comment]
你可以贴一下容器启动日志和容器内跑`debug.sh`的调试信息看看。

**Reply by kkkgo** [reply]
还有`docker stats`命令可以看资源占用/限制

**Reply by dudt** [reply]
root@debian:~# 
root@debian:~# docker logs paopaodns
=====PaoPaoDNS docker start=====
images build time : 2023-10-05 16:09:23 UTC
New version install ! Try clean...
ulimit adbove 1000000.
              total        used        free      shared  buff/cache   available
Mem:           7955         261        7490           0         204        7467
Swap:             0           0           0
              total        used        free      shared  buff/cache   available
Mem:           7.8G      261.4M        7.3G      500.0K      203.9M        7.3G
Swap:             0           0           0
MEMSIZE:7467
====ENV TEST====
MEM:500m 1000m 500000 1500mb
prefPC:100
CORES:-12-
POWCORES:-8-
ulimit :-1048576-
FDLIM :-4096-
TZ:-Asia/Shanghai-
UPDATE:-daily-
DNS_SERVERNAME:-Paopao.DNS-
SERVER_IP:-10.0.0.99-
ETHIP:-172.17.0.2-
DNSPORT:-53-
SOCKS5:-10.0.0.2:7891-
CNAUTO:-yes-
IPV6:-no-
CNFALL:-yes-
CUSTOM_FORWARD:-no-
AUTO_FORWARD:-no-
AUTO_FORWARD_CHECK:-no-
USE_MARK_DATA:-yes-
RULES_TTL:-600-
CN_TRACKER:-yes-
USE_HOSTS:-no-
HTTP_FILE:-no-
SAFEMODE:--
QUERY_TIME:-2000ms-
ADDINFO:-yes-
PLATFORM:-Linux e6c08d7ea401 6.1.0-13-cloud-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 Linux-
====ENV TEST====
Apply trackerslist...
load_mark_data
dat size pass.
1203940+0 records in
1203940+0 records out
1203940 bytes (1.1MB) copied, 3.242342 seconds, 362.6KB/s
1024+0 records in
1024+0 records out
1024 bytes (1.0KB) copied, 0.002940 seconds, 340.1KB/s
global_mark hash: OK.
/data
2023-10-20T02:46:33.447+0800	INFO	working directory changed	{"path": "/tmp"}
PID   USER     TIME  COMMAND
2023-10-20T02:46:33.449+0800	INFO	main config loaded	{"file": "mosdns.yaml"}
    1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh
   22 root      0:00 crond
  155 root      0:00 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.toml
  156 root      0:00 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt_socks.toml
  157 root      0:00 unbound -c /tmp/unbound_forward.conf -p
  158 root      0:00 mosdns start -d /tmp -c mosdns.yaml
  168 root      0:00 unbound -c /tmp/unbound_raw.conf -p
  171 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
  172 root      0:00 {data_update.sh} /bin/sh /usr/sbin/data_update.sh
  176 root      0:00 sleep 67
  184 root      0:00 ps
  185 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
191:C 20 Oct 2023 02:46:33.455 # WARNING Memory overcommit must be enabled! Without it, a background save or replication may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
191:C 20 Oct 2023 02:46:33.455 # systemd supervision requested or auto-detected, but Redis is compiled without libsystemd support!
Setting up watches.
Watches established.
WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/main: Permission denied
WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/community: could not connect to server (check repositories file)
WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/main: could not connect to server (check repositories file)
WARNING: updating and opening https://mirrors.zju.edu.cn/alpine/edge/community: could not connect to server (check repositories file)
2023-10-20 02:47:58 CST
/etc/unbound/named.cache Same hash, skip update.
used_memory_human:932.62K
2023-10-20 02:47:59 CST
/usr/sbin/Country-only-cn-private.mmdb.xz diff sha256sum, update...
newsum:db347859ea5607fa9a535ac3a5a4fcddfc347db367b6a20fd86350bc4a37b347
oldsum:c955bd2fe611321ca7b02e28ed72988f056233924b5d73180c0b54c40d9248c2
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51636  100 51636    0     0   154k      0 --:--:-- --:--:-- --:--:--  155k
/tmp/Country-only-cn-private.mmdb Download OK.
/usr/sbin/Country-only-cn-private.mmdb.xz Update OK.
/data/Country-only-cn-private.mmdb MODIFY 
2023-10-20 02:48:04 CST
/usr/sbin/trackerslist.txt.xz diff sha256sum, update...
newsum:47348cc94d210bb5ebb192e85e85f3e48f786341e25bba55158f9e03470450f2
oldsum:649d8ddd99c770303498c0e475c69d1be378181eef2134db9a5c9a0c446c909d
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2332  100  2332    0     0   7244      0 --:--:-- --:--:-- --:--:--  7264
/tmp/trackerslist.txt.xz.download Download OK.
  158 root      0:00 mosdns start -d /tmp -c mosdns.yaml
  158 root      0:00 [mosdns]
  158 root      0:00 [mosdns]
killall: mosdns: no process killed
mosdns reload...
killall: mosdns: no process killed
/data/trackerslist.txt: Waiting to apply the update...
  388 root      0:00 mosdns start -d /tmp -c mosdns.yaml
Setting up watches.
Watches established.
/data/trackerslist.txt: Waiting to apply the update...
/usr/sbin/trackerslist.txt.xz Update OK.
/data/trackerslist.txt MODIFY 
2023-10-20 02:48:12 CST
/data/global_mark.dat diff sha256sum, update...
newsum:6d18c69e709dcf049bce9fbe9f31fa5ba65db8077df0dfcbf078dba9b24f747c
oldsum:00b3fd44696d11154fbb69d51221b359f9eaa1f92f81da0a4fc87a2969cf6aa7
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1176k  100 1176k    0     0  2711k      0 --:--:-- --:--:-- --:--:-- 2715k
/tmp/global_mark.dat.download Download OK.
Apply trackerslist...
  388 root      0:00 mosdns start -d /tmp -c mosdns.yaml
  388 root      0:00 [mosdns]
mosdns reload...
killall: mosdns: no process killed
/data/global_mark.dat: Waiting to apply the update...
  528 root      0:00 mosdns start -d /tmp -c mosdns.yaml
Setting up watches.
Watches established.
/data/global_mark.dat: Waiting to apply the update...
/data/global_mark.dat MODIFY 
/data/global_mark.dat Update OK.
load_mark_data
dat size pass.
1203208+0 records in
1203208+0 records out
1203208 bytes (1.1MB) copied, 3.238259 seconds, 362.9KB/s
1024+0 records in
1024+0 records out
1024 bytes (1.0KB) copied, 0.002986 seconds, 334.9KB/s
global_mark hash: OK.
/data
  528 root      0:00 mosdns start -d /tmp -c mosdns.yaml
  528 root      0:00 [mosdns]
  528 root      0:00 [mosdns]
killall: mosdns: no process killed
mosdns reload...
killall: mosdns: no process killed
  647 root      0:00 mosdns start -d /tmp -c mosdns.yaml
Setting up watches.
Watches established.
/data/force_ttl_rules.txt MODIFY 
dnscrypt reload rules...
  647 root      0:00 mosdns start -d /tmp -c mosdns.yaml
  647 root      0:00 [mosdns]
  647 root      0:00 [mosdns]
killall: mosdns: no process killed
mosdns reload...
killall: mosdns: no process killed
  858 root      0:00 mosdns start -d /tmp -c mosdns.yaml
Setting up watches.
Watches established.
root@debian:~# docker exec -it e6c08d7ea401 /bin/sh
/data # debug.sh
=====PaoPaoDNS docker debug=====
[info] images build time : 2023-10-05 16:09:23 UTC
[info] ========== env info ==========
====ENV TEST====
MEM:500m 1000m 500000 1500mb
prefPC:100
CORES:-12-
POWCORES:-8-
ulimit :-1048576-
FDLIM :-4096-
TZ:-Asia/Shanghai-
UPDATE:-daily-
DNS_SERVERNAME:-Paopao.DNS-
SERVER_IP:-10.0.0.99-
ETHIP:-172.17.0.2-
DNSPORT:-53-
SOCKS5:-10.0.0.2:7891-
CNAUTO:-yes-
IPV6:-no-
CNFALL:-yes-
CUSTOM_FORWARD:-no-
AUTO_FORWARD:-no-
AUTO_FORWARD_CHECK:-no-
USE_MARK_DATA:-yes-
RULES_TTL:-600-
CN_TRACKER:-yes-
USE_HOSTS:-no-
HTTP_FILE:-no-
SAFEMODE:--
QUERY_TIME:-2000ms-
ADDINFO:-yes-
PLATFORM:-Linux e6c08d7ea401 6.1.0-13-cloud-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 Linux-
====ENV TEST====
[info] ========== process info ==========
PID   USER     TIME  COMMAND
    1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh
   22 root      0:00 crond
  156 root      0:16 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt_socks.toml
  171 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
  191 root      0:42 redis-server unixsocket:/tmp/redis.sock
  208 root      0:40 unbound -c /tmp/unbound_raw.conf -p
  830 root      0:08 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.toml
  858 root      0:33 mosdns start -d /tmp -c mosdns.yaml
  909 root      0:00 inotifywait -e modify,delete /etc/unbound/named.cache /data/Country-only-cn-private.mmdb /data/force_cn_list.txt /data/force_nocn_list.txt /data/global_mark.dat /data/trackerslist.txt /data/force_ttl_rules.txt
  913 root      0:00 /bin/sh
 1158 root      0:00 /bin/sh
 1164 root      0:00 {debug.sh} /bin/sh /usr/sbin/debug.sh
 1169 root      0:00 ps -ef
[info] ========== cn list info ==========
domain:whoami.ds.akahelp.net
domain:whoami.03k.org
[info] ========== reids info ==========
used_memory_human:1.51M
used_memory_rss_human:4.37M
used_memory_peak_human:1.51M
total_system_memory_human:7.77G
used_memory_lua_human:31.00K
used_memory_vm_total_human:63.00K
used_memory_scripts_human:181B
maxmemory_human:1.46G
(integer) 1847
[test] IP test, you will see that all the following IPs are your public network exit IP !
[test] ========== IP TEST START ==========
CN IP URL:
122.195.169.141
122.195.169.141
------------------
Non-CN IP URL:
122.195.169.141
122.195.169.141
122.195.169.141
122.195.169.141
122.195.169.141
------------------
IP INFO:
122.195.169.141
CN,Nanjing,Jiangsu
ASN4837/China Unicom
HTTP/1.1 
curl/8.4.0
Asia/Shanghai Time: 10/20/2023, 2:24:19 PM
------------------
----mosdns whoami aka dig:
"ns" "122.195.169.141"
------------------
----local-unbound whoami aka dig:
"ns" "122.195.169.141"
------------------
----mosdns whoami 03k dig:
122.195.169.141
------------------
----local-unbound whoami 03k dig:
122.195.169.141
[test] ========== IP TEST END ==========

[test] The DNS hijacking test, you will see timed out message !
[test] ========== DNS HIJACK START ==========
;; communications error to 9.8.7.6#53: timed out
;; communications error to 9.8.7.6#53: timed out
;; communications error to 9.8.7.6#53: timed out
;; no servers could be reached


[test] ========== DNS HIJACK END ==========
[test] CN domain test, you will see that the DNS resolution result is CN IP !
[test] ========== CN DOMAIN TEST START ==========
----mosdns CN dig:
www.taobao.com.danuoyi.tbcache.com.
119.167.220.175
119.167.220.174
------------------
----local-unbound CN dig:
www.taobao.com.danuoyi.tbcache.com.
119.167.220.174
119.167.220.175
[test] ========== CN DOMAIN TEST END ==========
[test] Non-CN domain test, you will see that the DNS resolution result is correct IP !
[test] ========== Non-CN DOMAIN TEST START ==========
----mosdns Non-CN dig:
youtube-ui.l.google.com.
172.217.31.174
142.250.207.46
------------------
----dnscrypt-unbound NOCN dig:
;; communications error to 127.0.0.1#5304: connection refused
;; communications error to 127.0.0.1#5304: connection refused
;; communications error to 127.0.0.1#5304: connection refused
------------------
----dnscrypt Non-CN dig:
youtube-ui.l.google.com.
142.251.220.46
142.251.220.78
------------------
----dnscrypt-socks5 Non-CN dig:
youtube-ui.l.google.com.
172.217.31.174
142.250.207.46
[test] ========== Non-CN DOMAIN TEST END ==========
[test] IPv6 Dual CN test: you will see that IPv6 is OK !
[test] ========== IPV6 CN DOMAIN TEST START ==========
[test] ========== IPV6 CN DOMAIN TEST END ==========
[test] IPv6 Dual Non-CN test: you will see that IPv6 is empty !
[test] ========== IPV6 Non-CN DOMAIN TEST START ==========
[test] ========== IPV6 Non-CN DOMAIN TEST END ==========
[test] IPv6 only Non-CN test: you will see that IPv6 is ok !
[test] ========== IPV6 ONLY Non-CN DOMAIN TEST START ==========
[test] ========== IPV6 ONLY Non-CN DOMAIN TEST END ==========

[info] ALL TEST FINISH.

启动日志和debug.sh，这样是正常的嘛。总感觉卡卡的，有时候这台服务器会内存直接沾满导致ssh连不进去只能重启解决，一般家用给多少内存呀？

**Reply by kkkgo** [reply]
从调试信息可以看出，是因为你CPU核心数太多了，所以内存占用也多了几倍。
后续我加个限制CPU核心数的选项吧。
docker可以限制一下可用内存。
家用给2~3G就可以了一般。

**Reply by kkkgo** [reply]
该功能已在新版本实现，详细参考更新日志。https://github.com/kkkgo/PaoPaoDNS/discussions/61
